Integrate Bedrock Asset System into the Core

aembler opened the pull request:
Supersedes https://github.com/concrete5/concrete5/pull/8448

Leaving this open pending review. (Note: we are committed to this approach, but I do want some review on this. We've already had good review by about scopes, what this means for backward compatibility, etc...)

KorvinSzanto commented on concrete/src/Feature/UsesFeatureInterface.php:
```diff
@@ -0,0 +1,14 @@
+
+namespace Concrete\Core\Feature;
+
+interface UsesFeatureInterface
```
This could use a docbloc that makes it clear why you'd use this

KorvinSzanto commented on concrete/src/Feature/UsesFeatureInterface.php:
```diff
@@ -0,0 +1,14 @@
+{
+
+    /**
+     * @return string[] Feature handles
```
Adding a description to this and an example of output: `return [Feature::SOME_FEATURE]` would make it easier to implement without needing to reference and existing implementation

KorvinSzanto commented on concrete/src/Feature/UsesFeatureInterface.php:
```diff
@@ -0,0 +1,14 @@
+    /**
+     * @return string[] Feature handles
+     */
+    public function getRequiredFeatures() : array;
```
PSR-12 requires `foo(): type;`

KorvinSzanto commented on concrete/src/Gathering/Item/Template/Template.php:
```diff
@@ -196,7 +194,7 @@ public function getGatheringItemTemplateSlotWidth(GatheringItem $item)
         $w = 0;
         $handles = $this->getGatheringItemTemplateFeatureHandles();
-        $assignments = \Concrete\Core\Feature\Assignment\GatheringItemAssignment::getList($item);
+        $assignments = GatheringItemAssignment::getList($item);
```
Did `GatheringItemAssignment` move? It looks like it won't properly resolve since the file was deleted and `use` was removed

aembler replied:
Yes, I'm going to fully strip these files out in another pull request. The feature and functionality are going to be fully removed from the core.

KorvinSzanto commented on concrete/src/Page/Theme/BedrockThemeTrait.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+namespace Concrete\Core\Page\Theme;
+
+trait BedrockThemeTrait
```
This trait could use a docblock at the top and for all members making it clear why you'd use this and what these methods do

KorvinSzanto commented on concrete/src/Page/Theme/BedrockThemeTrait.php:
```diff
@@ -0,0 +1,18 @@
+        $this->requireAsset('jquery');
+    }
+
+    public function getThemeGridFrameworkHandle()
```
In PHP 7.2 we can actually add return types in extended methods safely if we want to: https://3v4l.org/01O57

KorvinSzanto commented on concrete/themes/concrete/page_theme.php:
```diff
@@ -1,11 +1,11 @@
-        $this->requireAsset('jquery');
-    }
+
+    use BedrockThemeTrait;
```
Slick

aembler commented on concrete/src/Feature/Features.php:
```diff
@@ -0,0 +1,130 @@
+
+namespace Concrete\Core\Feature;
+
+class Features
```
Good idea, I've added one.

aembler commented:
@KorvinSzanto I've updated the PR per your review. Updates are in this commit: https://github.com/concrete5/concrete5/pull/8451/commits/e4fee4df942393c022751393e1b2fe285120c284

Additionally, I've added another commit that I hope should get the tests in this repo back to passing, so let's not merge until tests fully pass. I'll keep monitoring this morning.
